### ISSUE TITLE
Use here documents to declare manifests in the documention

### ISSUE THREAD

**Issue by gajus** [kubernetes/website#1558]
**This is a...** 
- [x] Feature Request
- [ ] Bug Report

**Problem:**

Documentation is hard to follow when reader needs to constantly switch between documentation and analysing the manifests in his IDE/git repository/wherever else the example links.

**Proposed Solution:**

Use [here documents](https://en.wikipedia.org/wiki/Here_document) to inline the manifests.

Take [Persistent Volume Walkthrough](http://kubernetes.io/docs/user-guide/persistent-volumes/walkthrough/) as an example.

> ## Using your claim as a volume
> 
> Claims are used as volumes in pods. Kubernetes uses the claim to look up its bound PV. The PV is then exposed to the pod.
> 
> ```
> $ kubectl create -f docs/user-guide/persistent-volumes/simpletest/pod.yaml
> 
> $ kubectl get pods
> NAME      READY     STATUS    RESTARTS   AGE
> mypod     1/1       Running   0          1h
> 
> $ kubectl create -f docs/user-guide/persistent-volumes/simpletest/service.json
> $ kubectl get services
> NAME              CLUSTER_IP       EXTERNAL_IP       PORT(S)       SELECTOR           AGE
> frontendservice   10.0.0.241       <none>            3000/TCP      name=frontendhttp  1d
> kubernetes        10.0.0.2         <none>            443/TCP       <none>             2d
> ```

To understand what you are actually doing I will need to pull the repository and read `docs/user-guide/persistent-volumes/simpletest/pod.yaml` and `docs/user-guide/persistent-volumes/simpletest/service.json` separately as I go through the examples.

It would be so much easier if this example read:

> ## Using your claim as a volume
> 
> Claims are used as volumes in pods. Kubernetes uses the claim to look up its bound PV. The PV is then exposed to the pod.
> 
> ```
> $ cat <<'EOF' | kubectl create -f -
> kind: Pod
> apiVersion: v1
> metadata:
>   name: mypod
>   labels:
>     name: frontendhttp
> spec:
>   containers:
>     - name: myfrontend
>       image: nginx
>       ports:
>         - containerPort: 80
>           name: "http-server"
>       volumeMounts:
>       - mountPath: "/usr/share/nginx/html"
>         name: mypd
>   volumes:
>     - name: mypd
>       persistentVolumeClaim:
>        claimName: myclaim-1
> EOF
> 
> $ kubectl get pods
> NAME      READY     STATUS    RESTARTS   AGE
> mypod     1/1       Running   0          1h
> 
> $ cat <<'EOF' | kubectl create -f -
> {
>   "kind": "Service",
>   "apiVersion": "v1",
>   "metadata": {
>     "name": "frontendservice"
>   },
>   "spec": {
>     "ports": [
>       {
>         "protocol": "TCP",
>         "port": 3000,
>         "targetPort": "http-server"
>       }
>     ],
>     "selector": {
>       "name": "frontendhttp"
>     }
>   }
> }
> EOF
> $ kubectl get services
> NAME              CLUSTER_IP       EXTERNAL_IP       PORT(S)       SELECTOR           AGE
> frontendservice   10.0.0.241       <none>            3000/TCP      name=frontendhttp  1d
> kubernetes        10.0.0.2         <none>            443/TCP       <none>             2d
> ```

**Page to Update:**

http://kubernetes.io/docs/user-guide/

**Considerations:**
- Inline manifest length
  
  There should be a sane limit that restricts inline manifest length, e.g. 100 lines. In cases where manifest is longer than this limit, use URL to the resource (instead of a relative file path) to refer to the manifest.
  
  The benefit of using a URL is that:
  - It enables to quickly open the resource in the browser. This reduces switching between different applications, assuming user is reading the documentation in a browser.
  - It enables to copy-paste the examples without needing to pull the repository containing the referred resources.
- Keeping the examples up-to-date
  
  Having the example manifests in the version control enables easy revision/ tracking of/ making changes.
  
  Therefore, documentation source pages should not inline manifests (also, respecting the length limit by hand is error-prone).
  
  Instead, documentation should always refer to the manifest by URL and there needs to be a build step that pulls the manifest,
  validates that the resource URL returns response 200, and depending on the length of the response, keeps the URL as it is
  or inlines the manifest in the output document.

**Comment by gajus** [comment]
cc @chrislovecnm

**Comment by chrislovecnm** [comment]
Another option is hosting them on http, which may be slmpler

kubectl create -f https://git.io/my-example.yaml

Woul that help or is embedded the yaml better?

**Comment by chrislovecnm** [comment]
Our have both?? You can see the doc and then use the url?

**Comment by gajus** [comment]
> Our have both?? You can see the doc and then use the url?

I have added "Considerations" section to the end of the original issue.

**Comment by chrislovecnm** [comment]
This makes a lot of sense to me @kubernetes/docs any feedback?

**Comment by PI-Victor** [comment]
wanted to open an issue about this, but i'm glad there's one that addresses it currently. the docs are confusing, they assume i want to clone the docs repo locally to be able to use an example.
bump for interest!

**Comment by tengqm** [comment]
There is currently syntax for embedding yaml manifests in the markdown documents, for example:

```
{% include code.html language="yaml" file="podpreset-preset.yaml" ghlink="/docs/tasks/inject-data-application/podpreset-preset.yaml" %}
```
This syntax will have the yaml file embedded in the webpage while at the same time provide a link for readers to download and try the example.
With https://github.com/kubernetes/website/pull/6546 merged, we can enforce syntax checking when new manifests are added.

Please feel free to reopen this if it is not solving the issue.
